### ISSUE TITLE
[Discuss] CosmosDb Provider additional options 

### ISSUE THREAD

**Issue by andreujuanc** [lucabriguglia/Kledex#56]
Hello!,

For a new project we are going to use CosmosDb provider, which works fine, but we realized that it does not use partition keys, or allow to change other options like throughput, etc .

I'll be submitting a PR for that, but while I get the code together i would love to get some input on this.

Two main ways of doing this: First is to add properties to DomainDbConfiguration, the second one is to allow the user to set the DocumentCollection, RequestOptions for both the database and the collection. The Database class is not needed.

Beside how we'll do it, of all the settings, I in particular, need these properties:

- ### PartitionKey
    I don't explicitly need to set it but I believe it should have been set to '/type' by default, and then subsequently used in the BaseDocumentRepository .

- ### OfferThroughput
    For development it'd be nice to set it to 400, since this eats the azure credits on my teams D:
    So, this property in particular would be nice to expose it in the settings class.

- ### ConsistencyLevel
     One of the key points of Cosmos is the different CL it can use. We should allow the dev to set it. 

At least these 3 options we should include, more than that then the developer should create its own ARM template to create the resource, **BUT**, the only problem is still the **PartitionKey**, which is must be set when doing read/write operations.

Thanks

**Comment by andreujuanc** [comment]
The CosmosDb SDK is almost here, should I wait for it ? I have some code implemented but not fully. Maybe it's worth the time?

**Comment by lucabriguglia** [comment]
It can always be updated later.

**Comment by xaviersolau** [comment]
Hello,
Is there any news about this issue?
Regards.

**Comment by lucabriguglia** [comment]
I'll work on it in the next few days.

**Comment by lucabriguglia** [comment]
Sorry, wrong branch name. That commit is obviously meant for a different feature.

**Comment by lucabriguglia** [comment]
So, I'm adding the RequestOptions directly to the DomainDbOptions:

```C#
public class DomainDbOptions
{
    public string DatabaseId { get; set; } = "DomainStore";
    public string AggregateCollectionId { get; set; } = "Aggregates";
    public string CommandCollectionId { get; set; } = "Commands";
    public string EventCollectionId { get; set; } = "Events";
    public RequestOptions RequestOptions { get; set; } = new RequestOptions();
}
```

Can be set in startup.cs:

```C#
services
    .AddKledex(options => 
    {
        options.PublishEvents = true;
        options.SaveCommandData = true;
        options.ValidateCommands = false;
    }, typeof(Product))
    .AddCosmosDbSqlProvider(Configuration, options =>
    {
        options.RequestOptions = new RequestOptions();
    })
    .AddServiceBusProvider()
    .AddFluentValidationProvider()
    .AddMemoryCacheProvider()
    .AddUI();
```

And will be used directly in the base document repository when creating a document:

```C#
protected BaseDocumentRepository(string collectionId, IDocumentClient documentClient, IOptions<DomainDbOptions> settings)
{
    _documentClient = documentClient;
    _databaseId = settings.Value.DatabaseId;
    _collectionId = collectionId;
    _requestOptions = settings.Value.RequestOptions;
}

public async Task<Document> CreateDocumentAsync(TDocument document)
{
    return await _documentClient.CreateDocumentAsync(GetUri(), document, _requestOptions);
}
```

**Comment by lucabriguglia** [comment]
This is the general idea, now we have manage the different scenarios: collections, database, queries, etc...

**Comment by lucabriguglia** [comment]
I have tried different options and at the end they get messy because there are too many combinations. I think the best way to approach this is to add the 3 suggested properties to the config class (PartitionKey, ConsistencyLevel and OfferThroughput) and if further customisation is needed I would say it's better to override the default implementation.

**Comment by xaviersolau** [comment]
well, I did kind of override on my side.... but I am not rally happy with this because this is not maintenance friendly in my opinion...

**Comment by lucabriguglia** [comment]
What would be your preferred approach?

**Comment by xaviersolau** [comment]
I will try to post a PR for that...

**Comment by lucabriguglia** [comment]
You can follow the progress on this branch: https://github.com/lucabriguglia/Kledex/tree/cosmos-db-options.

**Comment by lucabriguglia** [comment]
For the first iteration I have added the two options for `OfferThroughput` and `ConsistencyLevel` and default the `PartitionKey` to '/type'

**Comment by lucabriguglia** [comment]
Released in v2.4.
